packages/menu/src/MenuItem.ts

```diff
@@ -276,6 +324,108 @@ export class MenuItem extends LikeAnchor(Focusable) {
+            }, POINTERLEAVE_TIMEOUT);
+        }
+    }
+
+    protected handleSubmenuChange = (): void => {
+        this.click();
+    };
+
+    protected handleSubmenuPointerenter = (): void => {
+        if (this.leaveTimeout) {
+            clearTimeout(this.leaveTimeout);
+            delete this.leaveTimeout;
+        }
+    };
+
+    public async openOverlay({
+        immediate,
+    }: { immediate?: boolean } = {}): Promise<void> {
+        if (this.hasSubmenu && !this.open) {
+            this.open = true;
+            this.submenu = (
+                this.shadowRoot.querySelector(
+                    'slot[name="submenu"]'
+                ) as HTMLSlotElement
+            ).assignedElements()[0] as Menu;
+            this.submenu.addEventListener(
+                'pointerenter',
+                this.handleSubmenuPointerenter
+            );
+            this.submenu.addEventListener('change', this.handleSubmenuChange);
+            const popover = document.createElement('sp-popover');
+            popover.append(this.submenu);
+            this.submenu.tabIndex = 0;
+            this.submenu.removeAttribute('slot');
+            this.menuData.focusRoot?.submenuOpened(this);
+            const closeOverlay = openOverlay(this, 'click', popover, {
+                placement: this.isLTR ? 'right-start' : 'left-start',
+                receivesFocus: 'auto',
+                delayed: !immediate && false,
+            });
+            let closing = false;
+            this.closeOverlay = async (leave = false): Promise<void> => {
+                delete this.closeOverlay;
+                this.active = false;
+                if (this.submenu) {
+                    this.submenu.tabIndex = -1;
+                }
+                if (!leave) {
+                    closing = true;
+                }
+                (await closeOverlay)();
+            };
+            this.submenu.isSubmenu = this.closeOverlay;
+            const cleanup = (
+                event: CustomEvent<OverlayOpenCloseDetail>
+            ): void => {
+                if (this.submenu) {
+                    event.stopPropagation();
+                    this.submenu.slot = 'submenu';
+                    delete this.submenu.isSubmenu;
+                    this.append(this.submenu);
+                    this.open = false;
+                }
+                if (closing || event.detail.reason === 'external-click') {
+                    this.menuData.focusRoot?.dispatchEvent(
+                        new CustomEvent('close', {
+                            bubbles: true,
+                            composed: true,
+                            detail: { reason: 'external-click' },
+                        })
+                    );
+                }
+            };
+            this.addEventListener('sp-closed', cleanup as EventListener, {
+                once: true,
+            });
+        }
+    }
```

Comment:
Consider breaking `openOverlay` into smaller functions, if possible.

Reply:
Maybe in the future, but it closes over state for the time being.